src/Parser/DeathKnight/Unholy/Modules/Items/ColdHeartEfficiency.js

```diff
@@ -20,63 +20,117 @@ class ColdHeartEfficiency extends Analyzer {
+		this.concordanceStart = event.timestamp;
+	}
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = event.timestamp;
+	}
+  }
+  
   on_byPlayer_removebuff(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.UNHOLY_STRENGTH_BUFF.id) {
       this.unholyStrengthStart = 0;
     }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = 0;
+    }
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = 0;
+	}
   }
 
   on_byPlayer_applybuffstack(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.COLD_HEART_BUFF.id) {
       this.buffColdHeart = event.stack;
+	  if (this.buffColdHeart == 20) {
+		  this.timeAtMaxStacksStart = event.timestamp;
+	  }
     }
   }
 
   on_byPlayer_cast(event) {
     const spellId = event.ability.guid;
-    const unholyStrengthDuration = 15000;
-    const coldHeartMaxStack = 20;
-    const coldHeartOptimalMinStack = 17;
-    const remainingUnholyStrengthAllowed = 4000;
-    this.unholyStrengthRemaining = unholyStrengthDuration - (event.timestamp - this.unholyStrengthStart);
+	
     if (spellId === SPELLS.CHAINS_OF_ICE.id) {
-      this.totalColdHeartCasts++;
-      if (this.unholyStrengthRemaining > 0) {
-        if (this.unholyStrengthRemaining < remainingUnholyStrengthAllowed) {
-          if (this.buffColdHeart < coldHeartMaxStack) {
-            if (this.buffColdHeart >= coldHeartOptimalMinStack) {
-              this.correctColdHeartCasts++;
-            }
-          }
-        }
-      }
-      if (this.buffColdHeart === 20) {
-        this.correctColdHeartCasts++;
-      }
-      //This is only here for double casting Cold Heart. If Cold Heart is casted again before it reaches 2 stacks, the event won't update.
-      this.buffColdHeart = 0;
+		this.totalColdHeartCasts++;
+	  
+		this.timeAtMaxStacks = event.timestamp - this.timeAtMaxStacksStart;
+	
+		this.unholyStrengthRemaining = this.unholyStrengthDuration - (event.timestamp - this.unholyStrengthStart);
+		this.concordanceRemaining = this.concordanceDuration - (event.timestamp - this.concordanceStart);
+		this.khazgorothRemaining = this.khazgorothDuration - (event.timestamp - this.khazgorothStart);
+	  
+		//This checks wether or not any of the three buffs are about to fall off within the next 4 seconds.
+		if ((this.unholyStrengthRemaining > 0 && this.unholyStrengthRemaining < this.remainingDurationAllowed) || (this.concordanceRemaining > 0 && this.concordanceRemaining < this.remainingDurationAllowed) || (this.khazgorothRemaining > 0 && this.khazgorothRemaining < this.remainingDurationAllowed)){
+			if (this.buffColdHeart < this.coldHeartMaxStack && this.buffColdHeart >= this.coldHeartOptimalMinStack) {
+				this.correctColdHeartCasts++;
+			}else if (this.buffColdHeart < 20) {
+				this.castsTooEarly++;
+			}
+		}			
+		if (this.buffColdHeart === 20 && this.timeAtMaxStacks <= this.maxDurationAtMaxStacksAllowed) {
```

Comment:
While I understand by how you've built the if statements here that it shouldn't be possible to hit both this 'if' and the above 'if', you should make this explicit by making this an `else if`. I spent a few minutes trying to decipher how these blocks worked, and it would have been a lot easier with the explicit `else if`.